### PR TITLE
Initial constructor documentation overhaul

### PR DESCRIPTION
**Pull request by sffc**
This PR is based on the excellent feedback on https://github.com/unicode-org/icu4x/issues/2348#issuecomment-1249563656. I made the following improvements:

1. A brand new documentation page about constructors, which includes a decision chart similar to `std::collections`
2. Links to that documentation page from all constructors
3. Some editorializing in both the new documentation page and the main `icu_provider` page about why data providers are important and what explicit data providers achieve (I incorporated some text from Zibi's post)
4. Examples on each source data provider demonstrating how it works with its corresponding constructor (BlobDataProvider, StaticDataProvider, FsDataProvider, and AnyPayloadProvider)

I started by updating `icu_decimal::FixedDecimalFormatter`. If these changes look good, I'll do them to all the rest of the constructors.

What I did _not_ do in this PR is address concerns involving BakedDataProvider. That deserves its own PR, including a tutorial on how to leverage BakedDataProvider in your build.rs file.

### CODE REVIEW

**Comment by sffc** [comment]
Docs previews:

- [icu](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu/index.html)::[decimal](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu/decimal/index.html)::[FixedDecimalFormatter](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu/decimal/struct.FixedDecimalFormatter.html#) [[ Sample for what formatter docs can look like ]]
- [icu_decimal](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu_decimal/index.html)::[FixedDecimalFormatter](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu_decimal/struct.FixedDecimalFormatter.html#)
    - Observation: the links to `icu_provider::constructors` work from `icu_decimal` but not from `icu::decimal`?
- [icu_provider](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu_provider/index.html#) [[ Added additional paragraph at the top ]]
- [icu_provider](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu_provider/index.html)::[constructors](http://icu4x-pr-artifacts.storage.googleapis.com/gha/e510d7773e51d041cd17a0c5b2450e378a7ec0bc/docs/icu_provider/constructors/index.html#) [[ ALL NEW DOCS PAGE ]]

File: provider/core/src/helpers.rs [hunk 1]
```diff
@@ -187,6 +187,38 @@ fn test_hash_word_32() {
...
+            "`]"
+        )
+    };
+    (BUFFER, $krate:path, $see_also:path) => {
+        concat!(
+            "✨ **Enabled with the `\"serde\"` feature.**\n\n",
+            "Creates a new instance using a [`BufferProvider`](",
+            stringify!($krate),
+            "::BufferProvider).\n\n",
+            "[📚 Help choosing a constructor](",
```
Comment (robertbastian) [inline provider/core/src/helpers.rs:212]: I'd move this under the unstable link, i.e. mention the alternative, then explain how to choose between them, not the other way around.

File: provider/fs/src/fs_data_provider.rs [hunk 2]
```diff
@@ -11,13 +11,32 @@ use writeable::Writeable;
...
+///
+/// // Create a DataProvider from data files stored in a filesystem directory:
+/// let provider = FsDataProvider::try_new(concat!(
+///     env!("CARGO_MANIFEST_DIR"),
+///     "/tests/data/json",
+/// ))
+/// .expect("Directory exists");
+///
+/// // Check that it works:
+/// let formatter = HelloWorldFormatter::try_new_with_buffer_provider(
```
Comment (robertbastian) [inline provider/fs/src/fs_data_provider.rs:33]: Praise: Good idea to introduce HelloWorldFormatter.

**Review by robertbastian — DISMISSED** [review]

File: provider/core/src/constructors.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+//! the current or any previous version within the same SemVer major release. For example, if a
+//! data file is built to support FooFormatter version 1.1, then FooFormatter version 1.2 will be
+//! able to read the same data file. Likewise, backwards-compatible keys can always be included
+//! by `icu_datagen` to support older library versions.
+//!
+//! The `*_unstable` functions are only guaranteed to work on data built for the exact same version
+//! of ICU4X. The advantage of the `*_unstable` functions is that they result in the smallest code
+//! size and allow for automatic data slicing when `BakedDataProvider` is used. However, the type
+//! bounds of this function may change over time, breaking SemVer guarantees. These functions
+//! should therefore only be used when you have full control over your data lifecycle.
```
Comment (Manishearth) [inline provider/core/src/constructors.rs:74]: suggestion: "full control over your data lifecycle at compile time"?

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**